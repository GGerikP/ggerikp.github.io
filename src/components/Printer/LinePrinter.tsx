import React, { useCallback, useEffect, useRef, useState } from 'react';
import styled from 'styled-components';
import SegmentPrinter from './SegmentPrinter';

const PrintedLine = styled.div`
`;

interface LineSegment {
    text: string;
    link?: string;
    printDelayBefore?: number;
    printDelayAfter?: number;
}

export type Line = {
    lineSegments: LineSegment[];
}

type LinePrinterProps = {
    line: Line;
    lineIndex: number;
    typingSpeed: number;
    promptChars?: string;
    instantPrint: boolean;
    hasMoreLines: boolean;
    printNextLine?: () => void;
};

function LinePrinter({ line, lineIndex, typingSpeed, promptChars, instantPrint, hasMoreLines, printNextLine }: LinePrinterProps) {

    const sideEffectExecutedRef = useRef(false);
    const [segmentIndex, setSegmentIndex] = useState<number>(-1);
    const [displayedSegments, setDisplayedSegments] = useState<LineSegment[]>([]);

    // Called by the TextPrinter to print each next segment
    const printNextSegment = useCallback(() => {
        const updatedSegmentIndex = segmentIndex + 1;
        setSegmentIndex(updatedSegmentIndex);
        if (updatedSegmentIndex < line.lineSegments.length) {
            const nextSegment: LineSegment = {
                text: line?.lineSegments[updatedSegmentIndex]?.text,
                link: line?.lineSegments[updatedSegmentIndex]?.link,
                printDelayBefore: line?.lineSegments[updatedSegmentIndex]?.printDelayBefore,
                printDelayAfter: line?.lineSegments[updatedSegmentIndex]?.printDelayAfter,
            }
            const updatedDisplayedSegments = [...displayedSegments, nextSegment]
            setDisplayedSegments(updatedDisplayedSegments);
        } else {
            if (typeof printNextLine === 'function') {
                printNextLine();
            }
        }
    }, [segmentIndex, displayedSegments, line.lineSegments, printNextLine]);

    // Used to print the first segment
    useEffect(() => {
        if (!sideEffectExecutedRef.current) {
            printNextSegment();
        }
        sideEffectExecutedRef.current = true;
    }, [printNextSegment]);

    return (
        <PrintedLine id={`LinePrinter:${lineIndex}`}>
            {promptChars}
            {displayedSegments.map((segment, index) => {
                return (
                    <SegmentPrinter
                        key={index}
                        segment={segment}
                        segmentIndex={index}
                        typingSpeed={typingSpeed}
                        instantPrint={instantPrint}
                        hasMoreLines={hasMoreLines}
                        printNextSegment={printNextSegment}
                    />
                )
            })}
        </PrintedLine>
    );
}

export default LinePrinter;