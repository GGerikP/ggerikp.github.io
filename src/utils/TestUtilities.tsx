// Util Functions
export const checkIfCursorIsBlinking = (cursorElement: Element | null) => {
  const style = cursorElement ? getComputedStyle(cursorElement) : new CSSStyleDeclaration();
  expect(cursorElement).toBeInTheDocument();
  expect(cursorElement).toHaveStyle('display: inline-block;');
  expect(cursorElement).toHaveStyle('border-left: 2px solid white;');
  expect(style.getPropertyValue('animation')).toMatch(/.*\.5s step-end 0s infinite.*/);
};

export const checkIfCursorIsHidden = (cursorElement: Element | null) => {
  expect(cursorElement).toBeInTheDocument();
  expect(cursorElement).toHaveStyle('display: none;');
};

