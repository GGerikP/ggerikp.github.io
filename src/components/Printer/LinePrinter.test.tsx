import React from 'react';
import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react';
import LinePrinter, { Line } from './LinePrinter';
import { checkIfCursorIsBlinking, checkIfCursorIsHidden } from '../../utils/TestUtilities';

describe('LinePrinter', () => {

  it('renders correctly with no segments', async () => {
    const line: Line = { segments: [] };
    const renderedComponent = render(<LinePrinter
      id="line0"
      line={line}
      lineIndex={0}
      typingSpeed={0}
      instantPrint={true}
      isLastLine={false}
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      printNextLine={() => {}} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    await waitFor(async () => {
      expect(renderedComponent.container.querySelector('#line0')).toBeInTheDocument();
    }, { timeout: 1000 });
  });

  it('renders correctly with one segment where there are no more lines', async () => {
    const line: Line = { segments: [{ text: '' }] };
    const renderedComponent = render(<LinePrinter
      id="line0"
      line={line}
      lineIndex={0}
      typingSpeed={0}
      instantPrint={true}
      isLastLine={true} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    await waitFor(async () => {
      expect(renderedComponent.container.querySelector('#line0-segment0')).toBeInTheDocument();
      const cursorElement = renderedComponent.container.querySelector('#line0-segment0-cursor');
      expect(cursorElement).toBeInTheDocument();
      checkIfCursorIsBlinking(cursorElement);
    }, { timeout: 1000 });
  });

  it('renders correctly with two segments but there are no more lines', async () => {
    const line: Line = { segments: [{ text: '' }, { text: '' }] };
    const renderedComponent = render(<LinePrinter
      id="line0"
      line={line}
      lineIndex={0}
      typingSpeed={0}
      instantPrint={true}
      isLastLine={true} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    await waitFor(async () => {
      expect(renderedComponent.container.querySelector('#line0-segment0')).toBeInTheDocument();
      const cursorElement0 = renderedComponent.container.querySelector('#line0-segment0-cursor');
      expect(cursorElement0).toBeInTheDocument();
      checkIfCursorIsHidden(cursorElement0);
      const cursorElement1 = renderedComponent.container.querySelector('#line0-segment1-cursor');
      expect(cursorElement1).toBeInTheDocument();
      checkIfCursorIsBlinking(cursorElement1);
    }, { timeout: 1000 });
  });

  it('renders correctly with two segment where there are additional lines to follow', async () => {
    const line: Line = { segments: [{ text: '' }, { text: '' }] };
    const renderedComponent = render(<LinePrinter
      id="line0"
      line={line}
      lineIndex={0}
      typingSpeed={0}
      instantPrint={true}
      isLastLine={false} />);
    expect(renderedComponent.asFragment()).toMatchSnapshot();
    await waitFor(async () => {
      expect(renderedComponent.container.querySelector('#line0-segment0')).toBeInTheDocument();
      expect(renderedComponent.container.querySelector('#line0-segment0-cursor')).toBeInTheDocument();
      const cursorElement0 = renderedComponent.container.querySelector('#line0-segment0-cursor');
      expect(cursorElement0).toBeInTheDocument();
      checkIfCursorIsHidden(cursorElement0);
      const cursorElement1 = renderedComponent.container.querySelector('#line0-segment1-cursor');
      expect(cursorElement1).toBeInTheDocument();
      checkIfCursorIsHidden(cursorElement1);
    }, { timeout: 1000 });
  });

});
