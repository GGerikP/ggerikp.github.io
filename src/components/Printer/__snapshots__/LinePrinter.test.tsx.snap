// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinePrinter renders correctly with no segments 1`] = `
<DocumentFragment>
  <div
    class="sc-eDLKkx cfncZL"
    id="line0"
  >
    <span />
  </div>
</DocumentFragment>
`;

exports[`LinePrinter renders correctly with one segment where there are no more lines 1`] = `
<DocumentFragment>
  <div
    class="sc-eDLKkx cfncZL"
    id="line0"
  >
    <span />
    <span
      class="sc-guDLey wuijx"
      id="line0-segment0"
    >
      <span
        class="sc-hLQSwg gcsUJF"
      />
      <div
        class="sc-beySPh brsoYh"
        id="line0-segment0-cursor"
      >
         
      </div>
    </span>
  </div>
</DocumentFragment>
`;

exports[`LinePrinter renders correctly with two segment where there are additional lines to follow 1`] = `
<DocumentFragment>
  <div
    class="sc-eDLKkx cfncZL"
    id="line0"
  >
    <span />
    <span
      class="sc-guDLey wuijx"
      id="line0-segment0"
    >
      <span
        class="sc-hLQSwg gcsUJF"
      />
      <div
        class="sc-beySPh brsoYh"
        id="line0-segment0-cursor"
      >
         
      </div>
    </span>
  </div>
</DocumentFragment>
`;

exports[`LinePrinter renders correctly with two segments but there are no more lines 1`] = `
<DocumentFragment>
  <div
    class="sc-eDLKkx cfncZL"
    id="line0"
  >
    <span />
    <span
      class="sc-guDLey wuijx"
      id="line0-segment0"
    >
      <span
        class="sc-hLQSwg gcsUJF"
      />
      <div
        class="sc-beySPh brsoYh"
        id="line0-segment0-cursor"
      >
         
      </div>
    </span>
  </div>
</DocumentFragment>
`;
